misc/python/materialize/scratch.py

```diff
@@ -352,67 +274,30 @@ def launch_cluster(
         for d in descs
     ]
 
-    # Generate temporary ssh key for running commands remotely
-    tmpdir = tempfile.TemporaryDirectory()
-
-    identity_file = f"{tmpdir.name}/id_rsa"
-    spawn.runv(["ssh-keygen", "-t", "rsa", "-N", "", "-f", identity_file])
-    with open(f"{tmpdir.name}/id_rsa.pub") as pk:
-        local_pub_key = pk.read().strip()
-
     loop = asyncio.get_event_loop()
     loop.run_until_complete(
         asyncio.gather(
             *(
-                setup(
-                    i,
-                    subnet_id,
-                    local_pub_key,
-                    identity_file,
-                    (git_rev if d.checkout else "HEAD"),
-                )
+                setup(i, git_rev if d.checkout else "HEAD")
                 for (i, d) in zip(instances, descs)
             )
         )
     )
+
     hosts_str = "".join(
         (f"{i.private_ip_address}\t{d.name}\n" for (i, d) in zip(instances, descs))
     )
-
-    env = [f"{k}={v}" for k, v in extra_env.items()]
-
     for i in instances:
-        ssh.runv(
-            [f"echo {shlex.quote(hosts_str)} | sudo tee -a /etc/hosts"],
-            "ubuntu",
-            i.public_ip_address,
-            identity_file=identity_file,
-        )
+        mssh(i.instance_id, "sudo tee -a /etc/hosts", stdin=hosts_str.encode())
 
+    env = " ".join(f"{k}={shlex.quote(v)}" for k, v in extra_env.items())
     for (i, d) in zip(instances, descs):
         if d.launch_script:
-            ssh.runv(
-                [
-                    "cd",
-                    "~/materialize",
-                    ";",
-                ]
-                + env
-                + [
-                    "nohup",
-                    "bash",
-                    "-c",
-                    shlex.quote(d.launch_script),
-                    ">~/mzscratch-startup.out",
-                    "2>~/mzscratch-startup.err",
-                    "&",
-                ],
-                "ubuntu",
-                i.public_ip_address,
-                identity_file=identity_file,
+            mssh(
+                i.instance_id,
+                f"cd materialize && {env} nohup bash -c {shlex.quote(d.launch_script)} &> mzscratch.log &",
```

Comment:
single log for stderr and stdout is much better!

Reply:
Yeah, otherwise you lose the relative ordering of stdout and stderr and it's very confusing!